
<h1 style="color: #9f43c3ff; border-bottom: 3px solid #9f43c3ff; padding-bottom: 8px;">

AI & ML Course with BinX Week 3 Day 2 : Linear Regression & Model Evaluation

</h1>

<blockquote style="border-left: 3px solid #2e1a9aff; padding-left: 12px; margin-left: 0;">

<b>Day 2 Learning Objectives:</b>
- 📂 Load house price dataset and clean non-numeric attributes (`street`, `city`, `statezip`, `country`, `date`).
- 🎯 Define Feature Matrix ($X$) and Target Vector ($y$).
- 🤖 Train a <code>LinearRegression</code> model on an 80/20 train-test split.
- 📊 Extract feature coefficients and identify the strongest predictor.
- 📐 Evaluate model performance using MAE, RMSE, and $R^2$.
- 🎯 Benchmark model RMSE against a naive mean baseline to determine value addition.

</blockquote>

## <span style="color: #39b64eff">2.1 Environment Setup & Library Imports</span>

<blockquote style="border-left: 3px solid #FD1D1D; padding-left: 12px; margin-left: 0;">

We begin by importing core numerical, data manipulation, visualization, and Scikit-Learn modeling and evaluation metrics.

</blockquote>

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

df = pd.read_csv("../../Data/raw/HousePriceDataset.csv")

print("✅ Libraries successfully imported, and dataset is loaded.")

✅ Libraries successfully imported, and dataset is loaded.


## <span style="color: #309c42ff">2.2 Dataset Loading & Exploration</span>

<blockquote style="border-left: 3px solid #FD1D1D; padding-left: 12px; margin-left: 0;">

We load the housing dataset from <code>../../Data/raw/HousePriceDataset.csv</code> and inspect dataset dimensions, data types, and initial records.

</blockquote>

In [2]:
print(f"Dataset Shape: {df.shape[0]} rows, {df.shape[1]} columns\n")

df.info()

Dataset Shape: 4600 rows, 18 columns

<class 'pandas.DataFrame'>
RangeIndex: 4600 entries, 0 to 4599
Data columns (total 18 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   date           4600 non-null   str    
 1   price          4600 non-null   float64
 2   bedrooms       4600 non-null   float64
 3   bathrooms      4600 non-null   float64
 4   sqft_living    4600 non-null   int64  
 5   sqft_lot       4600 non-null   int64  
 6   floors         4600 non-null   float64
 7   waterfront     4600 non-null   int64  
 8   view           4600 non-null   int64  
 9   condition      4600 non-null   int64  
 10  sqft_above     4600 non-null   int64  
 11  sqft_basement  4600 non-null   int64  
 12  yr_built       4600 non-null   int64  
 13  yr_renovated   4600 non-null   int64  
 14  street         4600 non-null   str    
 15  city           4600 non-null   str    
 16  statezip       4600 non-null   str    
 17  country        4600 non-n

## <span style="color: #309c42ff">2.3 Data Cleaning & Feature Selection</span>

<blockquote style="border-left: 3px solid #FD1D1D; padding-left: 12px; margin-left: 0;">

We remove non-numeric string columns (such as street address, city, statezip, country, and date) to keep numerical predictors suitable for linear regression modeling.

</blockquote>

In [3]:
df['address'] = df['street'] + ", " + df['city'] + ", " + df['statezip'] + ", " + df['country']
df = df.drop(columns=['street', 'city', 'statezip', 'country'])
df = df.drop(columns=['address', 'date'])

## <span style="color: #309c42ff">2.4 Feature Matrix (x) & Target Vector (y) Separation</span>

<blockquote style="border-left: 3px solid #FD1D1D; padding-left: 12px; margin-left: 0;">

In Scikit-Learn conventions:
- <b>Feature Matrix ($X$):</b> A 2D DataFrame containing numerical predictor attributes.
- <b>Target Vector ($y$):</b> A 1D Series containing ground-truth price values (<code>price</code>).

</blockquote>

In [4]:
target_col = "price"

x = df.drop(columns=[target_col])
y = df[target_col]

print(f"Feature Matrix (X) Shape: {x.shape}")
print(f"Target Vector (y) Shape: {y.shape}")

Feature Matrix (X) Shape: (4600, 12)
Target Vector (y) Shape: (4600,)


## <span style="color: #309c42ff">2.5 Linear Regression Model Training</span>

<blockquote style="border-left: 3px solid #FD1D1D; padding-left: 12px; margin-left: 0;">

We partition the dataset into 80% training data and 20% testing data (`random_state=42`), initialize `LinearRegression`, fit the model on the training set, and generate predictions for the test set.

</blockquote>

In [5]:
x_train, x_test, y_train, y_test = train_test_split(
    x, y, test_size=0.2, random_state=42
)

model = LinearRegression()
model.fit(x_train, y_train)
predictions = model.predict(x_test)

print("✅ Model Successfully Trained!")

✅ Model Successfully Trained!


## <span style="color: #309c42ff">2.6 Model Coefficients & Strongest Predictor</span>

<blockquote style="border-left: 3px solid #FD1D1D; padding-left: 12px; margin-left: 0;">

We extract model coefficients (`model.coef_`) to assess feature influence and sort them by absolute magnitude to identify the dominant predictor.

</blockquote>

In [8]:
coefficients_df = pd.DataFrame({
    'Feature': x.columns,            
    'Coefficient': model.coef_,
    'Absolute_Effect': np.abs(model.coef_)
}).sort_values(by='Absolute_Effect', ascending=False)

print(coefficients_df)


          Feature    Coefficient  Absolute_Effect
5      waterfront  382459.666353    382459.666353
4          floors   69824.740108     69824.740108
0        bedrooms  -64497.461587     64497.461587
6            view   44755.841775     44755.841775
1       bathrooms   36520.440676     36520.440676
7       condition   29335.539392     29335.539392
10       yr_built   -2569.163533      2569.163533
2     sqft_living     186.049845       186.049845
8      sqft_above      96.860817        96.860817
9   sqft_basement      89.189028        89.189028
11   yr_renovated       8.259917         8.259917
3        sqft_lot      -0.514414         0.514414


## <span style="color: #309c42ff">2.7 Model Evaluation & Naive Baseline Comparison</span>

<blockquote style="border-left: 3px solid #FD1D1D; padding-left: 12px; margin-left: 0;">

We evaluate the model using MAE, RMSE, and $R^2$, and compare the model RMSE against a naive baseline predicting the average price (`y_train.mean()`).

</blockquote>

In [ ]:
mean_prediction = np.full_like(y_test, y_train.mean())
baseline_rmse = np.sqrt(mean_squared_error(y_test, mean_prediction))
rmse = np.sqrt(mean_squared_error(y_test, predictions))
mae = mean_absolute_error(y_test, predictions)
r2 = r2_score(y_test, predictions)

print("--- Baseline Performance ---")
print("Baseline RMSE: ", baseline_rmse)

print("\n--- Model Evaluation ---")
print("MAE (Mean Absolute Error): ",mae)
print("RMSE (Root Mean Squared Error): ",rmse)
print("R² (Coefficient of determination): ",r2)

if(rmse < baseline_rmse):
    result = baseline_rmse - rmse
    print("\n✅ Conclusion: The Linear Regression model ADDS VALUE over the mean baseline:\n ", result)
else:
    print("\n❌ Conclusion: The model DOES NOT add value over the baseline.")

--- Baseline Performance ---
Baseline RMSE:  1010500.7773258204

--- Model Evaluation ---
MAE (Mean Absolute Error):  210908.17325010512
RMSE (Root Mean Squared Error):  993439.3625461584
R² (Coefficient of determination):  0.032283856632784325

✅ Conclusion: The Linear Regression model ADDS VALUE over the mean baseline:
  17061.41477966204


## <span style="color: #309c42ff">2.8 Summary & Interpretation of Results</span>

<blockquote style="border-left: 3px solid #2e1a9aff; padding-left: 12px; margin-left: 0;">

📌 <b>Interpretation & Key Takeaways:</b>
- 🌊 <b>Strongest Predictor:</b> <code>waterfront</code> (+382,459.67) has the largest positive impact on house prices.
- 📈 <b>Top Drivers:</b> <code>floors</code> (+69,824.74), <code>view</code> (+44,755.84), and <code>bathrooms</code> (+36,520.44) strongly boost property valuation.
- 📊 <b>Performance Metrics:</b> Achieved MAE of ~$210,908.17, RMSE of ~$993,439.36, and $R^2$ of 0.0323.
- 🎯 <b>Baseline Benchmarking:</b> Baseline RMSE is ~$1,010,500.78. The Linear Regression model outperforms the naive baseline, reducing error by <b>$17,061.41</b> and proving it adds value.

</blockquote>